node/src/components/deploy_buffer.rs

```diff
@@ -232,47 +264,75 @@ impl DeployBuffer {
             .filter_map(|deploy_hash| {
-                self.collected_deploys
+                self.pending
                     .get(&deploy_hash)
                     .map(|deploy| (deploy_hash, deploy.clone()))
             })
             .collect();
-        self.collected_deploys
+        self.pending
             .retain(|deploy_hash, _| !deploy_map.contains_key(deploy_hash));
-        self.processed.insert(block, deploy_map);
+        self.proposed.insert(block, deploy_map);
     }
 
     /// Notifies the deploy buffer that a block has been finalized.
     fn finalized_block(&mut self, block: ProtoBlockHash) {
-        if let Some(deploys) = self.processed.remove(&block) {
-            self.collected_deploys
+        if let Some(deploys) = self.proposed.remove(&block) {
+            self.pending
                 .retain(|deploy_hash, _| !deploys.contains_key(deploy_hash));
             self.finalized.insert(block, deploys);
         } else if !block.is_empty() {
             // TODO: Events are not guaranteed to be handled in order, so this could happen!
-            error!("finalized block that hasn't been processed!");
+            error!("finalized block that hasn't been proposed!");
         }
     }
 
     /// Notifies the deploy buffer that a block has been orphaned.
     fn orphaned_block(&mut self, block: ProtoBlockHash) {
-        if let Some(deploys) = self.processed.remove(&block) {
-            self.collected_deploys.extend(deploys);
+        if let Some(deploys) = self.proposed.remove(&block) {
+            self.pending.extend(deploys);
         } else {
             // TODO: Events are not guaranteed to be handled in order, so this could happen!
-            error!("orphaned block that hasn't been processed!");
+            error!("orphaned block that hasn't been proposed!");
         }
     }
+
+    /// Prunes stale deploy information from the DeployBuffer, returns the total deploys pruned
+    fn prune(&mut self) -> usize {
+        fn prune_collection(map: &mut DeployCollection) -> usize {
+            let initial_len = map.len();
+            map.retain(|_hash, header| {
+                let now = Timestamp::now();
+                let lifespan = header.timestamp() + header.ttl();
+                lifespan > now
+            });
+            initial_len - map.len()
+        }
+        fn prune_proto_collection(proto_collection: &mut ProtoBlockCollection) -> usize {
+            let mut pruned = 0;
+            let mut remove = Vec::new();
+            for (proto_hash, proposed) in proto_collection.iter_mut() {
+                pruned += prune_collection(proposed);
+                if proposed.is_empty() {
+                    remove.push(*proto_hash);
+                }
+            }
+            proto_collection.retain(|k, _v| !remove.contains(&k));
+            pruned
+        }
+        let collected = prune_collection(&mut self.pending);
+        let proposed = prune_proto_collection(&mut self.proposed);
```

Comment:
Actually, maybe we could even prune `finalized` here: After all, it's only used for filtering `pending`, is it?

Reply:
True, it is used for filtering pending, but also for generating the list of deploys to include in a block - in my discussion with @goral09 he said that it was important not to remove from that list, since we are not accessing from disk the values it has seen in the past when we do either check.

Reply:
But if a deploy is expired, it will be removed from `pending`, so it'll never be considered for inclusion in a new block again, and won't need to be filtered?

…actually, we probably want to not only periodically filter those collections, but also refuse to add new deploys, if they have already expired, so that they aren't even in `pending` for 10 seconds?

Reply:
My argument for holding onto the deploy hashes in the `finalized` queue so that we don't re-include them (replay attack).

Reply:
> …actually, we probably want to not only periodically filter those collections, but also refuse to add new deploys, if they have already expired, so that they aren't even in `pending` for 10 seconds?

Sure, I'll add this check. @afck Similarly, you made a comment above:

>  I think we need some way to enforce an upper bound on the size of our collections.

How would we determine the size limit, and what should be the behaviour if we hit that limit?

Reply:
Once we have the persistent backend, we can prune the set of `finalized` hashes, replace it with a bloom filter and in case of it returning "I don't know" response we will consult the db.

Reply:
I think it would be good to just make sure we never include an expired deploy in `pending` again, and then also prune it from `finalized`, because `finalized` is the most problematic collection: it would otherwise grow indefinitely.

Reply:
> I think it would be good to just make sure we never include an expired deploy in `pending` again

but how would you do that if you want to prune `finalized`? 

I agree that `finalized` is the most problematic but we want to keep being correct and not sacrifice it in favour of limiting memory consumed.

Reply:
If a deploy is expired (`timestamp + ttl < now`), it will remain expired. So if at that point we remove it from both `pending` and `finalized`, and never add it to `pending` again (because `timestamp + ttl < now` — no need to consult `finalized`), we'll never propose it again.

Reply:
Yes but that's only a small subset of all deploys. It is not required that there's a TTL and it may as well be days/weeks/months long.

Reply:
Ah, I think I get it now… Obviously you didn't want to _clear_ the `finalized` queue but only _prune_ it from _expired_ deploys. Yes, that's safe. Sorry :zipper_mouth_face:

Reply:
> How would we determine the size limit, and what should be the behaviour if we hit that limit?

I don't know. In the long run, I imagine the user should be able to configure the size limit, and we should have some way to prioritize, so we can remove the lowest-priority deploys.

> It is not required that there's a TTL and it may as well be days/weeks/months long.

Sure — but that's not a reason _not_ to filter by TTL _if_ there is one?

Reply:
Spun off a call, is it ok to resolve this now?

Reply:
It is ok to implement it :sweat_smile: we can safely prune expired deploys from the `finalized` set _iff_ we check the expiration condition before adding incoming deploy to `pending`.